### PR TITLE
tp-qemu: Add some subtests when netkvm driver in use.

### PR DESCRIPTION
**Pull request by CongSmile**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1231015, 1230674, 1231007

Based on: autotest/tp-qemu#463

### CODE REVIEW

File: qemu/tests/netkvm_in_use.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            msg += "guest report driver is %s, " % output
+            msg += "expect is '%s'" % driver_name
+            raise error.TestFail(msg)
+
+    def bg_stress_test():
+        """
+        Run netperf_stress test as backgroud test
+        """
+        bg_stress_test = params.get("run_bgstress")
+        try:
```
Comment (xutian) [inline qemu/tests/netkvm_in_use.py]: Move try into if branch.

File: qemu/tests/netkvm_in_use.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        vm_wait_time = int(params.get("wait_before_subtest"))
+        logging.info("Sleep %ss before run the subtest", vm_wait_time)
+        time.sleep(vm_wait_time)
+        sub_type = params.get("sub_test")
+        if sub_type:
+            msg = "Run sub test %s during netkvm in use" % sub_type
+            error.context(msg, logging.info)
+            try:
+                utils_test.run_virt_sub_test(test, params, env, sub_type)
+            except Exception:
```
Comment (xutian) [inline qemu/tests/netkvm_in_use.py]: better to print original exception error info.

File: qemu/tests/netkvm_in_use.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                    utils_test.run_virt_sub_test, (test, params, env),
+                    {"sub_type": bg_stress_test})
+                stress_thread.start()
+                if not utils_misc.wait_for(lambda: env.get(bg_stress_run_flag),
+                                           wait_time, 0, 5,
+                                           "Wait %s test start" % bg_stress_test):
+                    raise error.TestError("Run stress test error")
+                return stress_thread
+        finally:
+            if bg_stress_test:
```
Comment (xutian) [inline qemu/tests/netkvm_in_use.py]: why set it to false here,  bg threads still running here

File: qemu/tests/netkvm_in_use.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        Check vm status after guest shutdown.
+        Expected result: ping should be successful between guest
+                         and host after guest reboot.
+        """
+        error.context("Check the nerkvm dirver after guest reboot",
+                      logging.info)
+        ping_test()
+
+    def ping_test():
+        bg_cmd = ping.run
```
Comment (xutian) [inline qemu/tests/netkvm_in_use.py]: where ping define?

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    wait_before_subtest = 100
+    ping_ext_host = "yes"
+    ext_host_get_cmd = "ip route | awk '/default/ { print $3 }'"
+    shell_client = "nc"
+    Windows:
+        netperf_server_link_win = "netserver-2.6.0.exe"
+        netperf_client_link_win = "netperf.exe"
+        target = ${netperf_client_link_win}
+        server_path_win = "c:\\"
+        client_path_win = "c:\\"
```
Comment (suqinhuang) [inline qemu/tests/cfg/netkvm_in_use.cfg]: Suggest to put Windows cfg under guest-os to keep style  unified

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    client_path = /var/tmp/
+    test_protocols = TCP_STREAM
+    netperf_client = ${vms}
+    netperf_server = localhost
+    netperf_test_duration = 200
+    wait_before_subtest = 100
+    ping_ext_host = "yes"
+    ext_host_get_cmd = "ip route | awk '/default/ { print $3 }'"
+    shell_client = "nc"
+    Windows:
```
Comment (suqinhuang) [inline qemu/tests/cfg/netkvm_in_use.cfg]: Hi coli,
This case is windows only, why need configure for Windows additional?

Comment (CongSmile) [inline qemu/tests/cfg/netkvm_in_use.cfg]: As the host need run netperf test.

**Comment by CongSmile** [comment]
Based on #443

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                devcon = "xcopy %s:\devcon\wnet_amd64\devcon.exe C: /Y"
+            i386:
+                devcon = "xcopy %s:\devcon\wnet_x86\devcon.exe C: /Y"
+        - with_hotplug:
+            sub_test = pci_hotplug
+            suppress_exception = False
+            pci_type = nic
+            pci_model = virtio-net-pci
+            match_string = "Virtio network device"
+            Host_RHEL.5:
```
Comment (suqinhuang) [inline qemu/tests/cfg/netkvm_in_use.cfg]: should be Host_RHEL.m5, as it won't be test on rhel5 host, we can ignore it.

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                pci_model = virtio
+            reference_cmd = wmic nic list
+            find_pci_cmd = wmic nic list
+            # Do nothing in Windows guest.
+            pci_test_cmd = echo PCI_NUM
+            run_dhclient = no
+            pci_num = 1
+            repeat_times = 100
+       - with_change_version:
+            sub_test = win_virtio_single_driver
```
Comment (suqinhuang) [inline qemu/tests/cfg/netkvm_in_use.cfg]: test have been updated to single_driver_update

**Comment by CongSmile** [comment]
@suqinhuang and @xiagao 

Could you help review this PR?

Thanks.

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    wait_before_subtest = 100
+    ping_ext_host = "yes"
+    ext_host_get_cmd = "ip route | awk '/default/ { print $3 }'"
+    shell_client = "nc"
+    Windows:
+        netperf_server_link_win = "netserver-2.6.0.exe"
+        netperf_client_link_win = "netperf.exe"
+        target_process = ${netperf_client_link_win}
+        server_path_win = "c:\\"
+        client_path_win = "c:\\"
```
Comment (suqinhuang) [inline qemu/tests/cfg/netkvm_in_use.cfg]: Let's put it in guest-os

**Comment by CongSmile** [comment]
@suqinhuang 

Updated. 
Could you help review it again?

Thanks.

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    hostpassword = redhat
+    login_timeout = 800
+    start_vm = yes
+    kill_vm_on_error = yes
+    check_guest_bsod = yes
+    driver_name = "netkvm"
+    run_bgstress = netperf_stress
+    bg_stress_run_flag = netperf_run
+    netperf_link = netperf-2.6.0.tar.bz2
+    server_path = /var/tmp/
```
Comment (spcui) [inline qemu/tests/cfg/netkvm_in_use.cfg]: this case only windows, need to update two lines?

Comment (CongSmile) [inline qemu/tests/cfg/netkvm_in_use.cfg]: The netperf cfg is for the host.
The guest windows part has been updated in the Windows.cfg.

Thanks.

**Comment by suqinhuang** [comment]
LGTM
Acked-by: Suqin Huang

**Comment by spcui** [comment]
Acked-by: scui@redhat.com

File: qemu/tests/cfg/netkvm_in_use.cfg [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        - with_hotplug:
+            sub_test = pci_hotplug
+            suppress_exception = False
+            pci_type = nic
+            pci_model = virtio-net-pci
+            match_string = "Virtio network device"
+            reference_cmd = wmic nic list
+            find_pci_cmd = wmic nic list
+            # Do nothing in Windows guest.
+            pci_test_cmd = echo PCI_NUM
```
Comment (xutian) [inline qemu/tests/cfg/netkvm_in_use.cfg]: looks very strange? why this command ?

Comment (CongSmile) [inline qemu/tests/cfg/netkvm_in_use.cfg]: This cmd is required by pci_hotplug.py which is used as a subtest.

**Comment by xutian** [comment]
looks ok for me. merged thanks @CongSmile

**Pull request merged**